File: Kotlin/Kotlin/src/week55/차트.java
package week55;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class 차트 {

    static int N, answer;
    static int[] percent, pick;
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = parseInt(br.readLine());
        percent = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        pick = new int[N];
        visited = new boolean[N];
        perm(0);
        System.out.println(answer);
    }

    private static void perm(int r) {
        if (r == N) {
            checkChart();
        }
        for (int i = 0; i < N; i++) {
            if (visited[i]) continue;
            visited[i] = true;
            pick[r] = percent[i];
            perm(r + 1);
            visited[i] = false;
        }
    }

    private static void checkChart() {
        int count = 0;
        int sum = 0;
        for (int i = 0; i < N; i++) {
            sum += pick[i];
            int total = sum;
            for (int j = 0; j < i; j++) {
                total -= pick[j];
                if (total < 50) break;
                if (total == 50) count++;
            }
        }
        answer = Math.max(answer, count);
    }

    private static int parseInt(String input) {
        return Integer.parseInt(input);
    }
}

File: Kotlin/Kotlin/src/week41/차트.kt
package week41

private lateinit var graph: List<Int>
private lateinit var visited: BooleanArray
private lateinit var pick: IntArray
private var n = 0
private var answer = 0
fun main() = with(System.`in`.bufferedReader()) {
    n = readLine().toInt()
    graph = readLine().split(" ").map { it.toInt() }
    pick = IntArray(n)
    visited = BooleanArray(n)
    val max =  graph.max()
    if (50 < max) {
        println(0)
    } else if (max == 50) {
        println(1)
    } else {
        permutation(0)
        println(answer)
    }
}

private fun permutation(r: Int) {
    if (r == n) {
        checkLine()
        return
    }
    for (i in 0 until n) {
        if (!visited[i]) {
            visited[i] = true
            pick[r] = graph[i]
            permutation(r + 1)
            visited[i] = false
        }
    }
}

private fun checkLine() {
    var count = 0
    for (i in pick.indices) {
        var sum = pick.slice(0..i).sum()
        for (j in 0 until i) {
            sum -= pick[j]
            if (sum == 50) {
                count++
                break
            }
            if (sum < 50) break
        }
    }
    answer = answer.coerceAtLeast(count)
}
